astropy/table/table.py

```diff
@@ -195,11 +203,28 @@ def as_array(self):
         if len(self.columns) == 0:
             return None
 
+        sys_byteorder = ('>', '<')[sys.byteorder == 'little']
+        native_order = ('=', sys_byteorder)
+
+        dtype = []
+
         cols = self.columns.values()
-        dtype = [descr(col) for col in cols]
+
+        for col in cols:
+            col_descr = descr(col)
+            byteorder = col.info.dtype.byteorder
+
+            if not keep_byteorder and byteorder not in native_order:
+                new_dt = np.dtype(col_descr[1]).newbyteorder('=')
+                col_descr = (col_descr[0], new_dt, col_descr[2])
+
+            dtype.append(col_descr)
+
         empty_init = ma.empty if self.masked else np.empty
         data = empty_init(len(self), dtype=dtype)
         for col in cols:
+            # This should automatically swap columns to their destination byte
```

Comment:
Ah, okay.  I was actually surprised myself when it worked that way.  Originally this had an explicit list of which columns to swap, and an explicit call to do that.  But then I realized that if you assign an array of the wrong byteorder to an existing recarray column then it will automatically swap.  Makes sense, but it seems a little surprising in this context.